tests/upgrades/test_activation_key.py

```diff
@@ -0,0 +1,143 @@
+        2. Upgrade the satellite to next/latest version
+        3. Update existing activation key with different entities
+        4. Postupgrade, Verify the activation key is intact, update and delete
+
+    :expectedresults: Activation key should create, update and delete successfully.
+    """
+    @classmethod
+    def setUpClass(cls):
+        cls.ak_name = 'preupgrade_activation_key'
+        cls.org_name = 'preupgrade_ak_org_name'
+        cls.cv_name = 'preupgrade_ak_cv'
+
+    @pre_upgrade
+    def test_pre_create_activation_key(self):
+        """Activation key with different entities are created
+
+        :id: preupgrade-a7443b54-eb2e-497b-8a50-92abeae01496
+
+        :steps:
+
+            1. In Preupgrade Satellite, Create activation key with different entities.
+
+        :expectedresults: Activation key should create successfully.
+        """
+        org = entities.Organization(name=self.org_name).create()
+        custom_repo = entities.Repository(
+            product=entities.Product(organization=org).create(),
+        ).create()
+        custom_repo.sync()
+        cv = entities.ContentView(
+            organization=org,
+            repository=[custom_repo.id],
+            name=self.cv_name
+        ).create()
+        cv.publish()
+        ak = entities.ActivationKey(content_view=cv, organization=org, name=self.ak_name).create()
+        org_subscriptions = entities.Subscription(organization=org).search()
+        for subscription in org_subscriptions:
+            provided_products_ids = [
+                prod.id for prod in subscription.read().provided_product]
+            if custom_repo.product.id in provided_products_ids:
+                ak.add_subscriptions(data={
+                    'quantity': 1,
+                    'subscription_id': subscription.id,
+                })
+        ak_subscriptions = ak.product_content()['results']
+        self.assertEqual(
+            {custom_repo.product.id},
+            {subscr['product']['id'] for subscr in ak_subscriptions}
+        )
+        ak.host_collection.append(entities.HostCollection().create())
+        ak.update(['host_collection'])
+        self.assertEqual(len(ak.host_collection), 1)
+
+    @post_upgrade
+    def test_post_crud_activation_key(self):
+        """Activation key is intact post upgrade and update/delete activation key
+
+        :id: postupgrade-a7443b54-eb2e-497b-8a50-92abeae01496
+
+        :steps:
+
+            1. Postupgrade, Verify activation key has same entities associated.
+            2. Update existing activation key with new entities
+            3. Delete activation key.
+
+        :expectedresults: Activation key should update and delete successfully.
+        """
+        org = entities.Organization().search(query={
+            'search': 'name={0}'.format(self.org_name)
+        })[0]
+        ak = entities.ActivationKey(organization=org).search(query={
+            'search': 'name={0}'.format(self.ak_name)
+        })[0]
```

Comment:
> Remove index based fetching - If it didnt find an activation key this will throw an 'IndexError: list index out of range'

If it didn't find an activation key, then it's a product bug :)

Reply:
Correct, but this way the test fail with python 'KeyError' error, we dont want show python errors if something is not found.

Instead we should shoe that ak is not migrated. The intention changes if you do so.